python/cudf/dataframe/columnops.py

```diff
@@ -334,16 +340,19 @@ def as_column(arbitrary, nan_as_null=True, dtype=None):
             data = as_column(pa.array([arbitrary]))
 
     elif isinstance(arbitrary, memoryview):
-        data = as_column(np.array(arbitrary))
+        data = as_column(np.array(arbitrary), dtype=dtype,
+                         nan_as_null=nan_as_null)
 
     else:
         try:
             data = as_column(memoryview(arbitrary))
         except TypeError:
             try:
-                data = as_column(pa.array(arbitrary))
+                data = as_column(pa.array(arbitrary, from_pandas=nan_as_null),
+                                 dtype=dtype)
             except pa.ArrowInvalid:
-                data = as_column(np.array(arbitrary))
+                data = as_column(np.array(arbitrary), dtype=dtype,
+                                 nan_as_null=nan_as_null)
```

Comment:
I was just reading the above original code and really curious about it. Your new changes are much better.

Reply:
😄